### PR TITLE
Supporting relationship queries

### PR DESCRIPTION
**Pull request by kitbs**
Hi there

Thanks for a great package. Hope you don't mind this massive PR out of the blue.

Inspired by #6, I have attempted to support querying for related models. It supports all simple Laravel relationship types (not polymorphic, yet), count queries, and nested relationships. There are also options to configure which relationships are queryable and countable.

I have used a different syntax than your examples in that issue, to make the parser more predictable. I have updated the readme in my branch so you can see all the syntax options possible, but here is how your examples would be expressed:

* `title: "My article" and has(comments) > 3`
* `title: "My article" and has(comments)` (equivalent to `has(comments) > 0`)
* `title: "My article" and has(comments { author: johndoe }) >= 1`
* `title: "My article" and has(category { name: Laravel })`  
  or `title: "My article" and category.name: Laravel` for simplicity
* `title: "My article" and has(category { name: Laravel or name: Vuejs })`

Basically everything inside the `{ ... }` is another expression scoped to the child model, and the `has` statements can be nested infinitely.

I have written extensive tests, but would really welcome scrutiny on what I've built. I really hope this is useful for you, and will happily take suggestions about how it can be improved.

Cheers!

### CODE REVIEW

**Comment by lorisleiva** [comment]
Hi there 👋

This is amazing! Thank you for taking the time to work on this massive feature.

I’m going to take the time to review this over the next couple of weeks but it’s definitely getting merged in!

I’ll keep you posted 🙂

**Comment by lorisleiva** [comment]
Hi there, 👋

I’ve had the chance to have a quick look at this PR this morning. Thanks again for the solid work!

There are just a few things that I’d like to adjust before releasing this.

- I’d really like to have a nice and clean DSL for this package. Unfortunately adding the `has` keyword and `{}` tokens makes that DSL more verbose and less intuitive. I’d like to stick to the original simpler language. This should be made possible by defining new less strict symbols that can later on be identified by visitors as relationships, normal where clauses or errors.
- Additionally, I think we should use the same column rules for relationships instead of creating yet another type of rule that the user needs to know about. Having a simple `relationship` Boolean on the column rules could do the trick and this could be dynamically deducted by checking the relationship methods.

Obviously you’ve already done a huge amount of work and I’m not expecting you to carry on putting more hours on this. I have no problem sorting this myself as soon as I can dedicate some time on this.

Thank again for the amazing work. Let me know if you want to dedicate some more time on this otherwise I’ll probably just merge it as-is and make the adjustments on following commits. 🙂

**Comment by alexjustesen** [comment]
Quick 2c from someone just following along...

Would it make sense to map the querying of the relationships to a specific field/fields so something like the following can be done as a clean syntax in the string?

Mapping example on a post model with a categories relationship
```php
$relationshipMap = [
    ['categories'] => ['name', 'slug'],
];
```

Query for the existence of a relationship:
```
has:categories
```

Query the fields set in the relationship map:
```
categories:laravel
```

Maybe go one step further and allow more complex searches using the tokens as mentioned in the initial PR.
```
has:categories(name:laravel)
```

I'm thinking along the lines of how Google does search operators for Gmail, figuring that's a pretty well known and simple syntax. [Search operators](https://support.google.com/mail/answer/7190?hl=en)

Just throwing out ideas and can't wait for this to be released!

**Comment by lorisleiva** [comment]
Hi there 👋 

I’m actually working on this at the moment and I’m planning on doing things a bit differently.

When you define a column as a relationship you’ll need to make sure the related Model also uses the search string traits and we’ll use this definition to create the sub queries.

On a side note, before working on this I’ve refactoring the entire parser into a compiler compiler, making grammar changes much much easier.

I’ll keep you posted.

**Comment by kitbs** [comment]
Hi @lorisleiva 

I hope there are some parts of this that you can make use of in your new solution. Sorry I won't be able to make any more time to help out.

Good luck with the refactor and I look forward to seeing how this turns out.

Cheers!

**Comment by lorisleiva** [comment]
Hi @kitbs 👋 

Thank you for your help with this!

As you probably noticed when writing this PR, it wasn't easy to stick to the original syntax described here https://github.com/lorisleiva/laravel-search-string/issues/6#issuecomment-525685294 whilst keeping the current parser.

For that reason, I had to refactor the entire parser and use an Llk compiler compiler ([from HOA](https://github.com/hoaproject/Compiler)) instead.

This allowed me to write the entire DSL syntax as a [grammar](https://github.com/lorisleiva/laravel-search-string/blob/ce9c4c612a5e0727d480efea12ffbe633d01891f/src/Compiler/Grammar.pp) using a predefined syntax. The grammar itself is then parsed to generate a compiler (hence the "compiler compiler").

With that in place, I could play with the grammar and write tests until I was happy with the result. In the end, I was able to reach 99% of the original syntax (`comments (title:foo)` was conflicting with implicit `AND` so it had to become `comments: (title:foo)`).

Anyway, the refactoring is now very much different so I can reuse anything from this PR but it certainly helped me a lot to get started and your tests and documentation changes have also helped me a lot.

Thank you again for your work and I hope you enjoy the new update that I'm thinking of releasing this weekend. 🙂
